File: Program's_Contributed_By_Contributors/C++ Programs/linkedlist/CircularLinkedList.cpp
#include <iostream>

class Node
{
public:
    int data;
    Node *next;

    Node(int value)
    {
        data = value;
        next = nullptr;
    }
};

class CircularLinkedList
{
private:
    Node *head;

public:
    CircularLinkedList()
    {
        head = nullptr;
    }

    ~CircularLinkedList()
    {
        if (head != nullptr)
        {
            Node *temp = head->next;
            while (temp != head)
            {
                Node *prev = temp;
                temp = temp->next;
                delete prev;
            }
            delete head;
        }
    }

    void append(int data)
    {
        Node *new_node = new Node(data);
        if (head == nullptr)
        {
            head = new_node;
            head->next = head;
        }
        else
        {
            Node *temp = head;
            while (temp->next != head)
            {
                temp = temp->next;
            }
            temp->next = new_node;
            new_node->next = head;
        }
    }

    void display()
    {
        if (head == nullptr)
        {
            std::cout << "Circular Linked List is empty." << std::endl;
            return;
        }

        Node *temp = head;
        do
        {
            std::cout << temp->data << " -> ";
            temp = temp->next;
        } while (temp != head);
        std::cout << head->data << std::endl;
    }
};

int main()
{
    CircularLinkedList cll;
    cll.append(1);
    cll.append(2);
    cll.append(3);
    cll.append(4);

    std::cout << "Circular Linked List:" << std::endl;
    cll.display();

    return 0;
}

File: circularlinkedlist.py
class Node:
   def __init__(self, my_data):
      self.data = my_data
      self.next = None

class circularLinkedList:  
   def __init__(self):
      self.head = None
   def add_data(self, my_data):
      ptr_1 = Node(my_data)
      temp = self.head    
      ptr_1.next = self.head

      if self.head is not None:
         while(temp.next != self.head):
            temp = temp.next
         temp.next = ptr_1
      else:
         ptr_1.next = ptr_1
      self.head = ptr_1

   def print_it(self):
      temp = self.head
      if self.head is not None:
         while(True):
            print("%d" %(temp.data)),
            temp = temp.next
            if (temp == self.head):
               break
my_list = circularLinkedList()
print("Elements are added to the list ")
my_list.add_data (56)
my_list.add_data (78)
my_list.add_data (12)
print("The data is : ")
my_list.print_it()
